src/main/daml/Daml/Finance/Account/Account.daml

```diff
@@ -23,38 +23,62 @@ template Account
     owner : Party
       -- ^ The account owner.
     controllers : Account.Controllers
-      -- ^ Controllers of transfers.
+      -- ^ Controllers of transfers and freezes.
     id : Id
       -- ^ Identifier of the account.
     description : Text
       -- ^ Description of the account.
     holdingFactoryCid : ContractId HoldingFactory.F
       -- ^ Associated holding factory.
+    unfreezers : Parties
+      -- ^ Parties who has frozen the account and are authorized to unfreeze it.
     observers : PartiesMap
       -- ^ Observers.
   where
-    signatory custodian, owner
+    signatory custodian, owner, unfreezers
     observer Disclosure.flattenObservers observers
 
     -- Outgoing controllers must be non-empty.
     ensure not . S.null $ controllers.outgoing
 
-    let account = AccountKey with custodian; owner; id
+    let
+      account = AccountKey with custodian; owner; id
+      mustNotBeFrozen = assertMsg @Update "account must not be frozen" $ S.null unfreezers
+      contextFrozen = "Frozen"
 
     interface instance Account.I for Account where
-      view = Account.View with custodian; id; owner; holdingFactoryCid; description; controllers
+      view = Account.View with
+        custodian; id; owner; holdingFactoryCid; description; controllers; unfreezers
       getKey = account
-      credit Account.Credit{quantity} =
+      credit Account.Credit{quantity} = do
+        mustNotBeFrozen
         exercise holdingFactoryCid HoldingFactory.Create with
           instrument = quantity.unit
           account
           amount = quantity.amount
           observers = M.empty
       debit Account.Debit{holdingCid} = do
+        mustNotBeFrozen
         vHolding <- view <$> fetch holdingCid
         assertMsg "accounts must match" $ vHolding.account == account
         exercise holdingFactoryCid HoldingFactory.Remove with
           actors = S.fromList [account.custodian, account.owner]; holdingCid
+      freeze Account.Freeze{freezers} = do
+        mustNotBeFrozen
+        assertMsg "Freezers must be non-empty" (not . S.null $ freezers)
+        newCid <- toInterfaceContractId @Disclosure.I <$> create this with unfreezers = freezers
+        -- adding freezers as observers to Account.R
+        coerceInterfaceContractId @Account.I <$> exercise newCid Disclosure.AddObservers with
+          disclosers = S.fromList (signatory this) <> freezers
+          observersToAdd = (contextFrozen, freezers)
+      unfreeze = do
+        newCid <- toInterfaceContractId <$> create this with unfreezers = S.empty
```

Comment:
After speaking with @markus-da today around locking, I'm thinking we should be more consistent between between the two as they are relatively close business wise.

For example, we should have the reason for `freezing` an account. Also, instead of relaying on the empty set to signify that an account isn't frozen, we could have something akin to what we have for `Locking` in `frozen : Optional Frozen` where `None` would signify its not frozen and `Some Frozen` would signify the account is frozen and for what reason.

Reply:
I changed to `frozen : Optional frozen`